lib/Dialect/ESI/runtime/cpp/include/esi/Ports.h

```diff
@@ -54,18 +55,59 @@ class WriteChannelPort : public ChannelPort {
 /// A ChannelPort which reads data from the accelerator.
 class ReadChannelPort : public ChannelPort {
 public:
-  using ChannelPort::ChannelPort;
+  ReadChannelPort(
+      const Type *type,
+      std::optional<std::chrono::microseconds> pollingInterval = std::nullopt)
+      : ChannelPort(type) {
+    if (pollingInterval)
+      this->pollingInterval = *pollingInterval;
+    else
+      this->pollingInterval = std::chrono::microseconds(100);
+  }
+
+  ~ReadChannelPort() { disconnect(); }
+
+  /// In addition to the basic 'connect()' behavior, this also start a
+  /// 'read' thread.
+  virtual void connect() override;
+
+  /// In addition to the basic 'disconnect()' behavior, this also stops the
+  /// 'read' thread and 'join's on it.
+  virtual void disconnect() override;
+
+  /// Asynchronous read.
+  virtual std::future<MessageData> readAsync() {
+    auto p = new std::promise<MessageData>();
+    if (!promiseQueue.push(p)) {
+      throw std::runtime_error("Channel is disconnected");
+      delete p;
+    }
+    return p->get_future();
+  }
+
+  /// Specify a buffer to read into. Blocking. Basic API, will likely change
+  /// for performance and functionality reasons.
+  virtual void read(MessageData &outData) {
```

Comment:
`bool read(MessageData &)` -> `void read(MessageData &outData)` API change makes this PR not compile when ESI runtime is built w/ the new callback service (used here: https://github.com/llvm/circt/pull/7153/files#diff-0432614eceb5e1876ca0ea5dbda0d10f9d36c4a5e0e66e8d7c96380bd618256eR116)

Reply:
Here's a fix i'm using internally:

```c++
  std::map<ReadChannelPort *, std::future<MessageData>> activeRead;

  while (!shutdown) {
    std::this_thread::sleep_for(std::chrono::microseconds(100));
    {
      std::lock_guard<std::mutex> g(listenerMutex);
      for (auto &[channel, cb] : listeners) {
        assert(channel && "Null channel in listener list");
        auto it = activeRead.find(channel);
        if (it == activeRead.end()) {
          activeRead[channel] = channel->readAsync();
        } else {
          auto &f = it->second;
          bool is_ready =
              f.wait_for(std::chrono::seconds(0)) == std::future_status::ready;
          if (!is_ready)
            continue;
          portUnlockWorkList.emplace_back(channel, cb, std::move(f.get()));
          activeRead.erase(it);
        }
      }
    }

...

Reply:
Yeah, there's definitely at least head of line blocking in the service thread of the other PR.